### PR TITLE
Network interface userspace access

### PR DESCRIPTION
**Pull request by jukkar**
Add new functions that can be used to add/lookup/remove IP address from the network interface when the application is running in userspace context. These functions are only available in userspace if CONFIG_NET_IF_USERSPACE_ACCESS is enabled.

### CODE REVIEW

**Comment by pfalcon** [comment]
Inconsistent naming of some functions is pointed by CI.

Beyond that, I'd suggest to explicitly mention in docstrings that interface index is 1-based (that's what we standardized on, right?). And even beyond that, I wonder if we should model such functions based on Linux' API (which would be based on classical BSD Unix API, even if this API is not mandated by POSIX).

**Comment by jukkar** [comment]
> I wonder if we should model such functions based on Linux' API (which would be based on classical BSD Unix API, even if this API is not mandated by POSIX).

What Linux APIs you are referring to?

**Comment by pfalcon** [comment]
> And even beyond that, I wonder if we should model such functions based on Linux' API (which would be based on classical BSD Unix API

Well, d'oh, that appears to be rather bloated and not directly fitting Zephyr model:

http://man7.org/linux/man-pages/man3/getifaddrs.3.html
http://man7.org/linux/man-pages/man3/if_nametoindex.3.html

Well, I'd still spend more time looking if something can be made "more consistent" with Linux ;-).

**Comment by jukkar** [comment]
The functions here can be used as a base if we want to add some BSD versions later.

File: subsys/net/ip/Kconfig [hunk 1]
```diff
@@ -529,6 +529,22 @@ config NET_DEFAULT_IF_CANBUS
 
 endchoice
 
+config NET_IF_USERSPACE_ADDR_ADD
```
Comment (tbursztyka) [inline subsys/net/ip/Kconfig]: Why having such atomic options? Couldn't we have a unique generic NET_IF_USERSPACE  instead?
If we expose more to the userspace following the same pattern, we will end up with a long list of Kconfig options.

Comment (jukkar) [reply subsys/net/ip/Kconfig]: I was thinking a bit more fine grained configuration options, but we can of course go with wider options too. I have no big preference here.

**Comment by jukkar** [comment]
Updated according to comments.

**Comment by jukkar** [comment]
recheck

**Comment by jukkar** [comment]
Rebased against latest upstream.

**Comment by jukkar** [comment]
* fixed unit tests
* added IPv4 netmask and gateway set functions so that they can be called from userspace

**Comment by jukkar** [comment]
Rebased against latest master.

**Review by tbursztyka — APPROVED** [review]

**Pull request merged**